pgmpy/factors/Factor.py

```diff
@@ -780,10 +780,12 @@ def __ne__(self, other):
         return not self.__eq__(other)
 
     def __hash__(self):
-        variables = sorted(self.variables)
+        variable_hashes = [hash(variable) for variable in self.variables]
+        sorted_var_hashes = list(sorted(variable_hashes))
         phi = self.copy()
         for axis in range(phi.values.ndim):
-            exchange_index = phi.variables.index(variables[axis])
+            exchange_index = variable_hashes.index(sorted_var_hashes[axis])
+            variable_hashes[axis], variable_hashes[exchange_index] = (variable_hashes[exchange_index], variable_hashes[axis])
             phi.variables[axis], phi.variables[exchange_index] = (phi.variables[exchange_index],
```

Comment:
@ankurankan Test are passing , first even I thought they should fail, because of the test case that were written by me in #578 but later I noticed line 789, hasn't been changes so we still are swapping the things in variables too, but this shouldn't have been done. @abhi95 in line 794 use hash values instead of variables and remove the line 789.